### ISSUE TITLE
Extending repository concepts

### ISSUE THREAD

**Issue by ftomassetti** [JetBrains/MPS-extensions#485]
To solve an issue in Modelix (https://github.com/modelix/modelix/issues/155) we should enrich the Module and Model concepts.

In practice, we would need to capture all the information present in .msd and .mps files within Module and Model concepts. In this way when we store Module and Model concepts in Modelix we would store all the information we need to later check out modules and models exactly equal to the original module and models we started from.

# Preliminary question

Are we ok enriching (or polluting) Module and Model with additional information or do we need for some reason to have them being more abstract/limited?

# Module

At this time there is information in msd files we do not save in the Module concept:
* moduleVersion (integer)
* compileInMPS (boolean)
* modelRoots (I am not sure we need those)
* facets
* sourcePath
* dependencies to modules
* dependencies to languages

## ModelRoots

I think most modules look like this:

```
    <modelRoot contentPath="${module}" type="default">
      <sourceRoot location="models" />
    </modelRoot>
```

Module using stubs may be different.    

## Facets

I think there is a limited number of facets, I would therefore think we could create an abstract ModuleFacet concept and then specific sub-concepts like JavaFacet which will have specific properties (generated and path)

```
    <facet type="java">
      <classes generated="true" path="${module}/classes_gen" />
    </facet>
```

## Source path

This seems to be empty in most modules:

```
<sourcePath />
```

Perhaps we could ignore it for now

## Dependencies

A typical dependency looks like this:
```
<dependency reexport="false">19fd8824-c1fd-40d2-b150-d0f433d4310e(Common)</dependency>
```

Here we specify only the direct dependencies, but we have another session in the module file where we store more dependencies. Probably that second list includes indirect dependencies and it is needed because we store the version of the dependency used. For example:

```
<module reference="19fd8824-c1fd-40d2-b150-d0f433d4310e(Common)" version="0" />
```

Perhaps we could have the concept "ModuleDependency" with a flag to indicate if it is an explicit/direct dependency or an implicit/indirect dependency.

## Language versions

They look like this:
```
<language slang="l:d4280a54-f6df-4383-aa41-d1b2bffa7eb1:com.mbeddr.core.base" version="5" />
```

One thing to notice is that the number of entries for languages is much longer than the list of languages displayed in the Used Languages tab of the module properties. I guess the list save on file contain also the indirect language dependencies. I assume we need this longer list as we need to save the version number for each of these languages.

# Model

At this time there is information in msd files we do not save in the Model concept:
* languages used
* Module imports
* the registry

## Language used

For example:
```
<use id="71934284-d7d1-45ee-a054-8c072591085f" name="org.iets3.core.expr.toplevel" version="4" />
```

## Imports

For example:
```
<import index="38vd" ref="r:51eaf3fa-4a2e-40ad-ad36-e94a8b928eb4(Common.Objects)" />
```

Here we probably do not care about the index, as it is useful only for the XML persistency

## Registry

For example:
```
    <language id="cfaa4966-b7d5-4b69-b66a-309a6e1a7290" name="org.iets3.core.expr.base">
      <concept id="5115872837156578546" name="org.iets3.core.expr.base.structure.PlusExpression" flags="ng" index="30dDZf" />
      <concept id="5115872837156576277" name="org.iets3.core.expr.base.structure.BinaryExpression" flags="ng" index="30dEsC">
        <child id="5115872837156576280" name="right" index="30dEs_" />
        <child id="5115872837156576278" name="left" index="30dEsF" />
      </concept>
    </language>
```

This whole concept of registry I think is useful only to define indexes used in the XML serialization, so I think we do not need to save it, unless I am missing something

**Comment by ftomassetti** [comment]
Regarding model imports. Currently, they are defined like this:
<img width="410" alt="Screenshot 2022-01-20 at 13 26 27" src="https://user-images.githubusercontent.com/439078/150338486-11408151-e8e5-41e0-bf5b-95af6e0bcc25.png">
<img width="451" alt="Screenshot 2022-01-20 at 13 26 32" src="https://user-images.githubusercontent.com/439078/150338470-f69a6101-2edb-45ce-a867-f683d1782b93.png">

I am not sure if we want to allow for references to models outside of Modelix, using some key instead of a reference to a model

**Comment by ftomassetti** [comment]
I created an initial PR around this as a basis for discussion

**Comment by slisson** [comment]
A model root for java stubs would require us to also store the jars on the server. In the future we may need to do that, but for now we can ignore the model roots.

The PR looks fine. We just have to implement the new properties in the adapters to fix the tests.

**Comment by ftomassetti** [comment]
Thank you for the feedback. I should have now implemented the adapters

**Comment by ftomassetti** [comment]
The PR has been merged, I think we can close this now

**Comment by ftomassetti** [comment]
I am reopening this as it seems it was merged already on 2021.1, so I need to backport this to 2020.3

**Comment by alexanderpann** [comment]
I am closing this issue because the backport is included in 2020.3.
